source/zipkin/tracer.cc

```diff
@@ -0,0 +1,137 @@
+#include <regex>
+
+#include "zipkin/util.h"
+#include "zipkin/tracer.h"
+#include "zipkin/zipkin_core_constants.h"
+
+namespace Zipkin {
+
+Span Tracer::startSpan(const std::string& span_name, uint64_t start_time) {
+  Span span;
+  Annotation cs;
+  std::string ip;
+  uint16_t port;
+  Endpoint ep;
+  uint64_t timestampMicro;
+
+  // Build the endpoint
+  getIPAndPort(address_, ip, port);
+  ep.setIpv4(ip);
+  ep.setPort(port);
+  ep.setServiceName(service_name_);
+
+  // Build the CS annotation
+  cs.setEndpoint(std::move(ep));
+  cs.setValue(ZipkinCoreConstants::CLIENT_SEND);
+
+  // Create an all-new span, with no parent id
+  span.setName(span_name);
+  uint64_t randonNumber = Util::generateRandom64();
+  span.setId(randonNumber);
+  span.setTraceId(randonNumber);
+  span.setStartTime(start_time);
+
+  // Set the timestamp globally for the span and also for the CS annotation
+  timestampMicro = Util::timeSinceEpochMicro();
+  cs.setTimestamp(timestampMicro);
+  span.setTimestamp(timestampMicro);
+
+  // Add CS annotation to the span
+  span.addAnnotation(std::move(cs));
+
+  span.setTracer(this);
+
+  return span;
+}
+
+Span Tracer::startSpan(const std::string& span_name, uint64_t start_time,
+                       SpanContext& previous_context) {
+  Span span;
+  Annotation annotation;
+  std::string ip;
+  uint16_t port;
+  Endpoint ep;
+  uint64_t timestampMicro;
+
+  // TODO We currently ignore the start_time to set the span/annotation timestamps
+  // Is start_time really needed?
+  timestampMicro = Util::timeSinceEpochMicro();
+
+  if ((previous_context.isSetAnnotation().sr) && (!previous_context.isSetAnnotation().cs)) {
+    // We need to create a new span that is a child of the previous span; no shared context
+
+    // Create a new span id
+    uint64_t randonNumber = Util::generateRandom64();
+    span.setId(randonNumber);
+
+    span.setName(span_name);
+
+    // Set the parent id to the id of the previous span
+    span.setParentId(previous_context.id());
+
+    // Set the CS annotation value
+    annotation.setValue(ZipkinCoreConstants::CLIENT_SEND);
+
+    // Set the timestamp globally for the span
+    span.setTimestamp(timestampMicro);
+  } else if ((previous_context.isSetAnnotation().cs) && (!previous_context.isSetAnnotation().sr)) {
+    // We need to create a new span that will share context with the previous span
+
+    // Initialize the shared context for the new span
+    span.setId(previous_context.id());
+    if (previous_context.parent_id()) {
+      span.setParentId(previous_context.parent_id());
+    }
+
+    // Set the SR annotation value
+    annotation.setValue(ZipkinCoreConstants::SERVER_RECV);
+  } else {
+    // Unexpected condition
+
+    // TODO Log an error
```

Comment:
You should have a stat here. If there are similar TODOs please also make sure there are error stats.